### PR TITLE
add Close for ActiveQueryTracker to close the file.

### PR DESCRIPTION
**Pull request by ianwoolf**
Signed-off-by: ianwoolf <btw515wolf2@gmail.com>

### CODE REVIEW

**Comment by ianwoolf** [comment]
Fix #10368

**Comment by roidelapluie** [comment]
Can we instead change the signature of getMMapedFile to `(filename string, filesize int, logger log.Logger) ([]byte, io.Closer, error)`?

**Comment by ianwoolf** [comment]
> Can we instead change the signature of getMMapedFile to `(filename string, filesize int, logger log.Logger) ([]byte, io.Closer, error)`?

emmm     looks more reasonable. Code has been updated

**Comment by LeviHarrison** [comment]
I think there are still windows errors happening in the test with this close function: https://github.com/prometheus/prometheus/pull/10321#issuecomment-1053031016, albeit a different error message. I'll try to see if I can figure out what's going on. We should hold off merging this until we have a fully working solution.

File: promql/query_logger.go [hunk 1]
```diff
@@ -197,3 +200,7 @@ func (tracker ActiveQueryTracker) Insert(ctx context.Context, query string) (int
 		return 0, ctx.Err()
 	}
 }
+
+func (tracker *ActiveQueryTracker) Close() {
+	tracker.closer.Close()
```
Comment (roidelapluie) [inline promql/query_logger.go:205]: We should check for nil here.

File: promql/query_logger_test.go [hunk 2]
```diff
@@ -111,7 +111,10 @@ func TestMMapFile(t *testing.T) {
 	filename := file.Name()
 	defer os.Remove(filename)
 
-	fileAsBytes, err := getMMapedFile(filename, 2, nil)
+	fileAsBytes, closer, err := getMMapedFile(filename, 2, nil)
+	if err != nil {
+		t.Cleanup(func() { closer.Close() })
```
Comment (roidelapluie) [inline promql/query_logger_test.go:116]:
```suggestion
		t.Cleanup(closer.Close)
```

Comment (ianwoolf) [reply promql/query_logger_test.go:116]: I wanted to write as this way at first. But `io.Closter` is `Close() error`, cannot use in `testing.Cleanup(f func())`

**Comment by LeviHarrison** [comment]
Ok I finally got my hands on a Windows computer to try out #10321... and I can't reproduce the test error. I think it's something to do with CircleCI, and because it looks like we're migrating to GitHub Actions soon, it's probably best to wait and see if the issue is fixed there.

**Comment by ianwoolf** [comment]
> #10368

Okay,  some PR have been blocked for a few months. hope that day will come sooner

**Comment by ianwoolf** [comment]
> Ok I finally got my hands on a Windows computer to try out #10321... and I can't reproduce the test error. I think it's something to do with CircleCI, and because it looks like we're migrating to GitHub Actions soon, it's probably best to wait and see if the issue is fixed there.

Okay, some PR have been blocked for a few months. hope that day will come sooner

**Comment by bboreham** [comment]
I have what looks like the same failure on GitHub Actions: https://github.com/prometheus/prometheus/runs/8251749473

File: promql/query_logger_test.go (hunk 2 above)
Comment (ianwoolf) [reply promql/query_logger_test.go:115]: ` cannot use closer.Close (value of type func() error) as func() value in argument to t.Cleanup`. so, it seems have to embase `closer.Close` into a function

**Comment by ianwoolf** [comment]
ping

**Comment by ianwoolf** [comment]
@roidelapluie   do you think there are any unresolved issues?   i need a review, please feel free to review    3ks

**Review by bboreham — APPROVED** [review]
Great stuff, thanks.

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**